### PR TITLE
[DML EP] Prefer MatMulInteger over MatMulIntegerToFloat in case of

### PR DESCRIPTION
**Pull request by raoanag**
### Description
Skip `MatMulIntegerToFloat` fusion in case of DML EP for cases where model uses Quantization before `MatMulInteger`. This is mainly done to be resource efficient, and we have better `MatMulInteger` Metacommand coverage which computes in int data type

### CODE REVIEW

**Review by PatriceVignola — DISMISSED** [review]

**Review by maggie1059 — APPROVED** [review]

File: onnxruntime/core/optimizer/matmul_integer_to_float.cc [hunk 1]
```diff
@@ -49,6 +49,49 @@ bool HasElementDataType(const NodeArg& node_arg, int32_t data_type) {
   return data_type == actual_data_type;
 }
 
+// Return total mnumber of Elements.
+static uint64_t NumElements(const TensorShapeProto* tensor_shape) {
+  if (nullptr == tensor_shape || tensor_shape->dim_size() < 1) {
+    return 0;
+  }
+  uint64_t num_elements = 1;
+  // First N-1 dimension must equal to 1
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/matmul_integer_to_float.cc:58]: Is this comment true for matmul?

File: onnxruntime/core/optimizer/matmul_integer_to_float.cc [hunk 2]
```diff
@@ -49,6 +49,49 @@ bool HasElementDataType(const NodeArg& node_arg, int32_t data_type) {
...
+  for (int i = 0; i <= tensor_shape->dim_size() - 1; i++) {
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/matmul_integer_to_float.cc:59]: `i <= tensor_shape->dim_size() - 1` is a bit weird, I would expect `i < tensor_shape->dim_size()` instead.

File: onnxruntime/core/optimizer/matmul_integer_to_float.cc [hunk 3]
```diff
@@ -49,6 +49,49 @@ bool HasElementDataType(const NodeArg& node_arg, int32_t data_type) {
...
+  return num_elements;
+}
+
+bool CheckMatMulLargeTensors(const Node& matmulinteger_node, const Node& cast_node) {
+  const auto a_def = matmulinteger_node.InputDefs()[0];
+  const auto b_def = matmulinteger_node.InputDefs()[1];
+  const int a_dim_size = a_def->Shape()->dim_size();
+  const int b_dim_size = b_def->Shape()->dim_size();
+  uint64_t inputA_bytes = NumElements(a_def->Shape());
+  uint64_t inputB_bytes = NumElements(b_def->Shape());
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/matmul_integer_to_float.cc:71]: Those are elements, not bytes

File: onnxruntime/core/optimizer/matmul_integer_to_float.cc [hunk 4]
```diff
@@ -49,6 +49,49 @@ bool HasElementDataType(const NodeArg& node_arg, int32_t data_type) {
...
+      }
+    }
+  }
+
+  int output_data_type = HasElementDataType(*cast_node.OutputDefs()[0], ONNX_NAMESPACE::TensorProto_DataType_FLOAT16) ? 2 : 4;
+  uint64_t total_bytes = (inputA_bytes + inputB_bytes) * output_data_type;
+
+  if (total_bytes > UINT32_MAX) {
+    return true;
+  }
```
Comment (PatriceVignola) [inline onnxruntime/core/optimizer/matmul_integer_to_float.cc:91]: Is it not possible to do within DirectML instead via a pass? This feels very implementation dependent. Depending on which path DirectML uses internally, this restriction wouldn't necessarily be needed, especially if we add fused shader implementations in the future for this op.

Comment (raoanag) [reply onnxruntime/core/optimizer/matmul_integer_to_float.cc]: This will be addressed in DirectML, added to backlog

**Review by PatriceVignola — APPROVED** [review]

**Pull request merged**